lib/scholar/naive_bayes/multinomial.ex

```diff
@@ -81,87 +84,201 @@ defmodule Scholar.NaiveBayes.Multinomial do
+      raise ArgumentError,
+            """
+            expected x to have shape {num_samples, num_features}, \
+            got tensor with shape: #{inspect(Nx.shape(x))}\
+            """
+    end
 
-    opts =
-      [
-        sample_weights_flag: opts[:sample_weights] != nil,
-        priors_flag: opts[:priors] != nil
-      ] ++ opts
+    if Nx.rank(y) != 1 do
+      raise ArgumentError,
+            """
+            expected y to have shape {num_samples}, \
+            got tensor with shape: #{inspect(Nx.shape(y))}\
+            """
+    end
 
-    x_type = to_float_type(x)
+    {num_samples, num_features} = Nx.shape(x)
 
-    {sample_weights, opts} = Keyword.pop(opts, :sample_weights, Nx.tensor(1.0, type: x_type))
-    sample_weights = Nx.tensor(sample_weights, type: x_type)
+    if num_samples != Nx.axis_size(y, 0) do
+      raise ArgumentError,
+            """
+            expected first dimension of x and y to be of same size, \
+            got: #{num_samples} and #{Nx.axis_size(y, 0)}\
+            """
+    end
+
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    x_type = to_float_type(x)
 
-    {priors, opts} = Keyword.pop(opts, :priors, Nx.tensor(0.0, type: x_type))
-    class_priors = Nx.tensor(priors)
     {alpha, opts} = Keyword.pop!(opts, :alpha)
     alpha = Nx.tensor(alpha, type: x_type)
 
-    fit_n(x, y, sample_weights, class_priors, alpha, opts)
+    if Nx.shape(alpha) not in [{}, {num_features}] do
+      raise ArgumentError,
+            """
+            when alpha is list it should have length equal to num_features = #{num_features}, \
+            got: #{Nx.size(alpha)}\
+            """
+    end
+
+    num_classes = opts[:num_classes]
+
+    priors_flag = opts[:class_priors] != nil
+
+    {class_priors, opts} = Keyword.pop(opts, :class_priors, :nan)
+    class_priors = Nx.tensor(class_priors)
+
+    if priors_flag and Nx.size(class_priors) != num_classes do
+      raise ArgumentError,
+            """
+            expected class_priors to be list of length num_classes = #{num_classes}, \
+            got: #{Nx.size(class_priors)}\
+            """
+    end
+
+    sample_weights_flag = opts[:sample_weights] != nil
+
+    {sample_weights, opts} = Keyword.pop(opts, :sample_weights, :nan)
+    sample_weights = Nx.tensor(sample_weights, type: x_type)
+
+    if sample_weights_flag and Nx.shape(sample_weights) != {num_samples} do
+      raise ArgumentError,
+            """
+            expected sample_weights to be list of length num_samples = #{num_samples}, \
+            got: #{Nx.size(sample_weights)}\
+            """
+    end
+
+    opts =
+      opts ++
+        [
+          priors_flag: priors_flag,
+          sample_weights_flag: sample_weights_flag
+        ]
+
+    fit_n(x, y, class_priors, sample_weights, alpha, opts)
+  end
+
+  defnp fit_n(x, y, class_priors, sample_weights, alpha, opts) do
+    # TODO: Why not just to_float_type?
+    x_type = Nx.Type.merge(to_float_type(x), {:f, 32})
```

Comment:
@msluszniak I am not sure why `Nx.Type.merge/2` is used here. Do you maybe remember?

Reply:
I can see this being the case if the algorithm just doesn't work with precision less than f32. My SVD impl in Nx for instance only works properly for f32, not even f64.

Reply:
Yes, that was exactly my intention.

Reply:
@msluszniak So, the algorithm doesn't work with `f16`? Any specific part?

Reply:
Actually, I tested the implementation and it works for `:bf16` and `:f16`, so it's rather over-zeal 😅. Feel free to delete these merges.

Reply:
Great, thanks!